### DISCUSSION TITLE
XGBoostLSSRegressor - Scikit Learn API

### DISCUSSION THREAD

**Discussion by joshdunnlime** [StatMixedML/XGBoostLSS#66] [ToDo]
Is there an appetite to add a scikit learn API for this? If so, very happy to help contrib. Also for LightGBMLSS.

**Comment by StatMixedML** [comment]
@joshdunnlime Thanks for the interest in the project.

An extension to scikit-learn or sktime is on the to-do, but I won't find the time for that. Very happy if you volunteer for this!

**Comment by joshdunnlime** [comment]
Amazing. I'm very happy to start looking into and hopefully start a PR. It would be good to get some input on naming conventions and over all design when I get there.

**Comment by StatMixedML** [comment]
> It would be good to get some input on naming conventions and over all design when I get there.

Sure let's align on this.

**Comment by kmedved** [comment]
I just want to add my support for this. This would dramatically increase the ease of use for this package.

**Comment by joshdunnlime** [comment]
I've made the first few commits and added the first PR #68.

Most of the code is taken from the XGBoost Sklearn package and adapted. I think it makes sense to align with that as closely as possible.

There are a few feature that I haven't preserved fully.
- Cupy support. I'm on a mac m1 so cannot test - nvidia have stopped supporting. I have no idea if this is something we wish to support in xgblss. This can easily be added in another version.
- Incremental/continuous training. Feels like much added complexity for a niche feature. Does this work with LSS training?
- validate_features. to be added.

**Comment by joshdunnlime** [comment]
Any comments or reviews on this? I plan to pick this up again this week

**Comment by StatMixedML** [comment]
Thanks for the PR. I am still on vacation and haven't had the time to look into it so far. Please continue.

**Comment by StatMixedML** [comment]
@joshdunnlime [skpro](https://github.com/sktime/skpro/issues/135) is a good motivation for having a scikit learn api.

**Reply by joshdunnlime** [reply]
For sure! There are many other use cases as well. This will certainly make it easier for the wider python DS community to get their hands dirty with!

**Comment by joshdunnlime** [comment]
Funnily enough, I scheduled some time this evening to pick this back up. I've now push a new PR. I think it is close to being ready to review and merge.

**Comment by joshdunnlime** [comment]
Additionally, there is an ongoing discussion on sklearn about how probabilistic predictions are made in an sklearn context. An agreed API here would be an excellent thing to adopt but for now I have followed XGBLSS's predict methodology. https://github.com/scikit-learn/scikit-learn/issues/23334

**Comment by fkiraly** [comment]
`skpro` proposes an API for probabilistic predictions and implements already an interface definition, with stringent tests.

It also integrates fully with `sktime`, the "scikit-learn-like" for time series forecasting.

Opened issue with suggestion on integration here: https://github.com/StatMixedML/XGBoostLSS/issues/69

**Reply by joshdunnlime** [reply]
+1 from me

**Comment by StatMixedML** [comment]
@joshdunnlime @fkiraly 

I have added separate branches to intergrate the project to the respective frameworks:

- [sklearn_integration](https://github.com/StatMixedML/XGBoostLSS/tree/sklearn_integration)
- [sktime_integration](https://github.com/StatMixedML/XGBoostLSS/tree/sktime_integration)
- [skpro_integration](https://github.com/StatMixedML/XGBoostLSS/tree/skpro_integration)

This allows us to have the different versions separate from the current master-branch. Once the integrations are done we can merge them to master.

@joshdunnlime Can you please move the PR to the new branch, thanks.

**Comment by joshdunnlime** [comment]
@StatMixedML my understanding is that simply adding the sklearn API will allow skpro to utilise xgblss (with some small changes on the skpro side). This in turn will allow sktime to use xgblss.

Most of the work is to add missing xgblss distributions in skpro/sktime side as detailed in #69. In the meantime, any not implemented could simply have a `NotImplemented` error. The other thing to consider here is pytorch vs tensorflow distributions and normalising flows in tensorflow.

In addition, the xgblss native predict api could be extended to accommodate the `predict_probab` etc methods.

I will shift my branch to the new branch and add a PR.
